src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,122 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/ast_construction.h"
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "antlr4-runtime.h"
+#include "src/common/logging/logging.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+
+namespace raksha::ir::auth_logic {
+
+static Principal ConstructPrincipal(
+    AuthLogicParser::PrincipalContext& context) {
+  return Principal(context.ID()->getText());
+}
+
+static datalog::Predicate ConstructPredicate(
+    AuthLogicParser::PredicateContext& context) {
+  std::vector<std::string> args = utils::MapIter<std::string>(context.pred_arg(), [](auto* statement) { return statement->ID()->getText(); });
+//   for (auto& statement : context.pred_arg()) {
+//     args.push_back(std::move(statement->ID()->getText()));
+//   }
+  // args =  utils::MapIter<std::string>(context->pred_arg(), [] );
+  datalog::Sign sign = context.NEG() == nullptr ? datalog::Sign::kPositive
+                                                : datalog::Sign::kNegated;
+  return datalog::Predicate(context.ID()->getText(), args, sign);
+}
+
+static BaseFact ConstructVerbphrase(
+    Principal left_principal, AuthLogicParser::VerbphraseContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::PredphraseContext*>(
+          &context)) {  // PredicateContext
+    return BaseFact(
+        Attribute(left_principal, ConstructPredicate(*ctx->predicate())));
+  }
+  auto* ctx = CHECK_NOTNULL(dynamic_cast<AuthLogicParser::ActsAsPhraseContext*>(
+      &context));  // ActsAsPhraseContext
+  return BaseFact(
+      CanActAs(left_principal, ConstructPrincipal(*ctx->principal())));
+}
+
+static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::PrinFactContext*>(&context)) {
+    return ConstructVerbphrase(ConstructPrincipal(*ctx->principal()),
+                               *ctx->verbphrase());
```

Comment:
Should the results of things like `principal()` and `verbphrase()` also go through `CHECK_NOTNULL` before being dereferenced?

This is more important in C++ that it is in languages like Java or Python. While Java and Python have a special `null` value that is checked by the underlying virtual machine, C++'s handling of `nullptr` is relatively crude: `nullptr` is, in most cases, literally the address `0`. On most machines, to support this, the virtual page mapper intentionally leaves the 0th page unmapped so that any reads or writes to it will immediately segfault.

But there are systems where this is not done or is not possible. For instance, the WINE emulator steals this 0th page for its own bookkeeping. Systems that do not have a virtual mapping layer often use this page, as it is some real, physical memory that would otherwise be dead. There exist cases where it is not safe, and in fact a real memory safety issue, to read or write to address 0. So we have to be careful with our handling of pointers that may be `nullptr`, and try to catch the `nullptr`s before they are dereferenced.

Reply:
Doesn't the fact that `if (auto *ctx =..)` true guarantee that `ctx` is not nullptr?

Reply:
It guarantees that `ctx` is not `nullptr`, but I don't think there's any guarantees on the result of `ctx->principal()` which is also being dereferenced via the `*`.

Reply:
aah.. missed that.